tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -224,6 +224,286 @@ static bool HasCanonicalDimensionNumbers(
   return true;
 }
 
+//===----------------------------------------------------------------------===//
+// mhlo.Einsum conversion patterns.
+//===----------------------------------------------------------------------===//
+
+// Looks through a set of dimension that has been marked as Reduction axes
+// if it is found within the set, then we set it as "reduction", otherwise
+// we can label it as "parallel"
+SmallVector<StringRef, 3> GetEinsumLoopsAttrs(
+    const llvm::SmallSetVector<StringRef, 4>& input_ind,
+    const llvm::SmallSetVector<StringRef, 4>& reduction_dims) {
+  SmallVector<StringRef, 3> res;
+  for (const StringRef& dim : input_ind) {
+    if (!reduction_dims.contains(dim)) {
+      res.push_back(getParallelIteratorTypeName());
+    } else {
+      res.push_back(getReductionIteratorTypeName());
+    }
+  }
+  return res;
+}
+
+// Check for where each dimension of the output operand came from
+// If that dimension is dynamically sized then we push it to dyn_size vector
+SmallVector<Value, 2> ExtractDynamicEinsumSizes(
+    OpBuilder& b, const Location& loc, const Value& lhs, const Value& rhs,
+    const SmallVector<std::string>& lhs_loop_vec,
+    const SmallVector<std::string>& rhs_loop_vec,
+    const SmallVector<std::string>& output_loop_vec) {
+  std::vector<Value> tensor_vec = {lhs, rhs};
+  std::vector<ArrayRef<int64_t>> tensor_shape_vec;
+  for (auto tensor_val : tensor_vec) {
+    auto tensor_current_type =
+        tensor_val.getType().dyn_cast<RankedTensorType>();
+    if (!tensor_current_type) return {};
+    tensor_shape_vec.push_back(tensor_current_type.getShape());
+  }
+  SmallVector<Value, 2> dyn_sizes;
+  for (const StringRef& dim_ind : output_loop_vec) {
+    int arg_number = 0;
+    auto dim_ind_it =
+        std::find(lhs_loop_vec.begin(), lhs_loop_vec.end(), dim_ind);
+    auto dim_ind_pos = dim_ind_it - lhs_loop_vec.begin();
+    if (dim_ind_it == lhs_loop_vec.end()) {
+      arg_number = 1;
+      dim_ind_it = std::find(rhs_loop_vec.begin(), rhs_loop_vec.end(), dim_ind);
+      dim_ind_pos = dim_ind_it - rhs_loop_vec.begin();
+    }
+    if (tensor_shape_vec[arg_number][dim_ind_pos] != ShapedType::kDynamicSize)
+      continue;
+    dyn_sizes.push_back(
+        b.create<memref::DimOp>(loc, tensor_vec[arg_number], dim_ind_pos));
```

Comment:
The `arg_number` makes things complicated to me. Sometimes you use `lhs/rhs` stuff and sometimes you use `arg_number`.

I think we can drop `arg_number` in this case because we already assumes that there are only two operands. I'm saying it because the function explicitly spell `lhs`/`rhs` in the function arguments. I think we can do something like

```cpp
    auto dim_ind_it =
        std::find(lhs_loop_vec.begin(), lhs_loop_vec.end(), dim_ind);
    Value dim_size;
    if (dim_ind_it != lhs_loop_vec.end()) {
      // query from lhs vars
    } else {
      // query from rhs vars
    }
    dyn_sizes.push_back(dim_size);
```

Reply:
done